### ISSUE TITLE
[a11y]: Modal: doesn't return focus properly on dismissal

### ISSUE THREAD

**Issue by finken2** [carbon-design-system/carbon#13680]
### Package

@carbon/react

### Browser

Chrome

### Operating System

MacOS

### Package version

v1.25.0

### React version

v16.12.0

### Automated testing tool and ruleset

manual testing

### Description

When launching a Modal from a button, focus goes inside the Modal, as it should. When dismissing the Modal, focus should return to the button where it was launched from. 
This issue was mentioned in #12727, but i thought it was worth splitting out separately. 

https://www.w3.org/WAI/WCAG21/Understanding/focus-order

Optimally, we'd automatically save off focus on launch and reset it on dismissal. Alternatively, having a prop for a selector to return focus too would also be an option.

### WCAG 2.1 Violation

WCAG 2.1, focus order 

### Reproduction/example

any implementation of Modal from a button

### Steps to reproduce

Launch a modal from a button. Dismiss modal. Note focus does not return to previous location. Focus is on the main `body` node after Modal dismissal.

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/carbon-design-system/carbon/blob/f555616971a03fd454c0f4daea184adf41fff05b/.github/CODE_OF_CONDUCT.md)
- [X] I checked the [current issues](https://github.com/carbon-design-system/carbon/issues) for duplicate problems

**Comment by tay1orjones** [comment]
Thanks for reporting this. I can reproduce it in [the storybook](https://react.carbondesignsystem.com/?path=/story/components-modal--with-state-manager) by focusing the "Launch modal" button, pressing <kbd>enter</kbd> to open the modal, and pressing <kbd>esc</kbd> to close the modal. Focus is not returned to the button

**Comment by tay1orjones** [comment]
How this was originally outlined in parent #12727 

> Whatever launches a dialog, when it is dismissed, that item should regain focus. In the implementations in React storybook, the focus is not shown when the dialog closes, and the user must shift+tab back to put the button in focus.

**Comment by tay1orjones** [comment]
When this is tackled, we should cover the solution in automated keyboard nav tests like we're doing in https://github.com/carbon-design-system/carbon/issues/13747 and https://github.com/carbon-design-system/carbon/issues/13748

**Comment by tw15egan** [comment]
I'm not sure we can handle this on our end, since `Modal` and `ComposedModal` are just toggled via the `open` prop which can be toggled by any element, in most cases a `button`. The `ComposedModal` has an example of how to handle this functionality, however:

```jsx
const closeButton = useRef();

<Button ref={closeButton} onClick={() => setOpen(true)}>
    Launch composed modal
</Button>


<ComposedModal
     open={open}
     onClose={() => {
     setOpen(false);
     setTimeout(() => {
         closeButton.current.focus();
    });
}}>
```

Any other ideas?

**Comment by tw15egan** [comment]
Put up https://github.com/carbon-design-system/carbon/pull/14355 that adds in a new prop `launcherButtonRef` that should allow for an easier configuration of this functionality.
